CamDetection.as

```diff
@@ -0,0 +1,125 @@
+#if TMNEXT
+
+void RenderMenuMain()
+{
+	if (!Setting_DebugShowActiveCamInMenu) {
+		return;
+	}
+	UI::TextDisabled("Active Cam: " + tostring(g_activeCameraType));
+}
+
+const Reflection::MwClassInfo@ tmTy = Reflection::GetType("CTrackMania");
+const Reflection::MwMemberInfo@ gameSceneMember = tmTy.GetMember("GameScene");
+const uint GameCameraNodOffset = gameSceneMember.Offset + 0x10;
+
+void UpdateActiveGameCam()
+{
+	g_activeCameraType = ActiveCam::None;
+
+	auto gameCameraNod = Dev::GetOffsetNod(GetApp(), GameCameraNodOffset);
+
+	// This is null in the menu and when cameras aren't being used.
+	if (gameCameraNod is null) {
+		return;
+	}
+
+	auto camModelNod = Dev::GetOffsetNod(gameCameraNod, 0x58);
+	auto camControlNod = Dev::GetOffsetNod(gameCameraNod, 0x68);
+	// 0x1 for std, 0x2 for alt.
+	bool cam1Alt = Dev::GetOffsetUint8(gameCameraNod, 0x24) == 0x2;
+	bool cam2Alt = Dev::GetOffsetUint8(gameCameraNod, 0x25) == 0x2;
+	bool cam3Alt = Dev::GetOffsetUint8(gameCameraNod, 0x26) == 0x2;
+
+	// Always 4 when backwards, and seemingly always 0 otherwise
+	bool isBackwards = Dev::GetOffsetUint32(gameCameraNod, 0xB0) == 0x4;
```

Comment:
Still not a fan of all these offsets tbh. Have you checked these offsets across game updates? How stable are they?

I'm afraid that Nadeo is going to release an update and make these offsets moot, and we will be forced to spend a long time on updating the offsets, or finding different methods of getting this information, or maybe we even need to remove the feature altogether.

This is my general view of using the Dev API btw, nothing specifically against this. I'm just wondering how stable it is, because it **will** add a lot of maintenance work. (Perhaps we need a checklist that we manually go through every game update.)